### ISSUE TITLE
Console output caused Orleans Worker Role to restart

### ISSUE THREAD

**Issue by DixonDs** [dotnet/orleans#2464]
We saw this in logs just before the role restarted:
```
[00000004] [11/30/2016 01:26:24.63] [INFO]  Substatus is  [11/30T01:26Z]Unhandled Exception: Probable I/O race condition detected while copying memory. The I/O package is not thread safe by default. In multithreaded applications, a stream must be accessed in a thread-safe way, such as a thread-safe wrapper returned by TextReader's or TextWriter's Synchronized methods. This also applies to classes like StreamWriter and StreamReader.
   at System.Buffer.InternalBlockCopy(Array src, Int32 srcOffsetBytes, Array dst, Int32 dstOffsetBytes, Int32 byteCount)
   at System.IO.StreamWriter.Write(Char[] buffer, Int32 index, Int32 count)
   at System.IO.TextWriter.SyncTextWriter.WriteLine(String value)
   at System.Console.WriteLine(String value)
   at Orleans.Runtime.ConsoleText.WriteLine(ConsoleColor color, String msg)
   at Orleans.Runtime.LoggerImpl.TrackException(Exception exception, IDictionary`2 properties, IDictionary`2 metrics)
   at Orleans.Runtime.LoggerImpl.WriteLogMessage(Int32 errorCode, Severity sev, String format, Object[] args, Exception exception)
   at Orleans.Runtime.LoggerImpl.Log(Int32 errorCode, Severity sev, String format, Object[] args, Exception exception)
   at Orleans.Runtime.LoggerExtensions.Warn(Logger logger, ErrorCode errorCode, String message, Exception exception)
   at Orleans.Runtime.SafeTimerBase.<HandleAsyncTaskTimerCallback>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

It seems it is related with the following issue: http://stackoverflow.com/questions/33915790/console-out-and-console-error-race-condition-error-in-a-windows-service-written 

Is it possible to update `ConsoleText` so that it does nothing when there is no console available?

**Comment by benjaminpetit** [comment]
I think Orleans should only use `ConsoleText` if a components is designed to output something to the console. If you set the `TraceToConsole` flag to false you should be fine. If this flag is set to false and there is still console output, then maybe this is a bug?

The only "illegal" usage that I found so far is here: https://github.com/dotnet/orleans/blob/master/src/Orleans/CodeGeneration/GrainInterfaceUtils.cs#L194
I think we should remove this occurrence.

**Comment by DixonDs** [comment]
@benjaminpetit Since `TraceToConsole` is defaulting to `true`, I think it is still an easy trap. I think it is better to detect if console is available and either set `TraceToConsole` default value based on that or change `ConsoleText` to be save to use with no console available.

**Comment by benjaminpetit** [comment]
Yes you are right, we could set `TraceToConsole` default value based on the availability of a console.

I know that in full .net we can use `Environment.UserInteractive` to detect that, but it doesn't seem available on .net Core. We could fallback to `true` as for now in this case.

**Comment by sergeybykov** [comment]
Resolved via #2501.
